sklearn/neighbors/tests/test_neighbors.py

```diff
@@ -1315,59 +1427,26 @@ def test_neighbors_badargs():
         nbrs.radius_neighbors_graph(X, mode="blah")
 
 
-# TODO: Remove filterwarnings in 1.3 when wminkowski is removed
 @pytest.mark.filterwarnings("ignore:WMinkowskiDistance:FutureWarning:sklearn")
-def test_neighbors_metrics(n_samples=20, n_features=3, n_query_pts=2, n_neighbors=5):
+@pytest.mark.parametrize("metric", COMMON_VALID_METRICS)
+def test_neighbors_metrics(
+    metric, n_samples=20, n_features=3, n_query_pts=2, n_neighbors=5
+):
     # Test computing the neighbors for various metrics
     # create a symmetric matrix
-    V = rng.rand(n_features, n_features)
-    VI = np.dot(V, V.T)
-
-    metrics = [
-        ("euclidean", {}),
-        ("manhattan", {}),
-        ("minkowski", dict(p=1)),
-        ("minkowski", dict(p=2)),
-        ("minkowski", dict(p=3)),
-        ("minkowski", dict(p=np.inf)),
-        ("chebyshev", {}),
-        ("seuclidean", dict(V=rng.rand(n_features))),
-        ("mahalanobis", dict(VI=VI)),
-        ("haversine", {}),
-    ]
-    if sp_version < parse_version("1.8.0.dev0"):
-        # TODO: remove once we no longer support scipy < 1.8.0.
-        # wminkowski was removed in scipy 1.8.0 but should work for previous
-        # versions.
-        metrics.append(
-            ("wminkowski", dict(p=3, w=rng.rand(n_features))),
-        )
-    else:
-        # Recent scipy versions accept weights in the Minkowski metric directly:
-        metrics.append(
-            ("minkowski", dict(p=3, w=rng.rand(n_features))),
-        )
-
     algorithms = ["brute", "ball_tree", "kd_tree"]
-    X = rng.rand(n_samples, n_features)
+    X_train = rng.rand(n_samples, n_features)
+    X_test = rng.rand(n_query_pts, n_features)
 
-    test = rng.rand(n_query_pts, n_features)
+    metric_params_list = _generate_test_params_for(metric, n_features)
 
-    for metric, metric_params in metrics:
+    for metric_params in metric_params_list:
+        # Some metric (e.g. Weighted minkowski) are not
+        # supported by KDTree
+        exclude_kd_tree = "minkowski" in metric and "w" in metric_params
```

Comment:
The previous comment would imply re-adding `metric not in neighbors.KDTree.valid_metrics` to define `exclude_kd_tree`.